### DISCUSSION TITLE
Questions about PoAs in scenario

### DISCUSSION THREAD

**Discussion by ruben-bc** [InterDigitalInc/AdvantEDGE#356] [Q&A] [answered]
Hi all again,

I'm doing a scenario where a terminal connects to a PoA while reproducing a streaming video and short time after changes to an another PoA without interrupting the reproduction. For more context, each PoA have to be connected to a different server and in the mobility event the second PoA be able to download the video if it's not locally yet. Also, the idea is do it through a RTMP server. I have a few questions:

1) How can each PoA connect to a streaming server of videos? For example retranstiming in streaming from saved videos in pc, for simplicity.

2) The MEC021 supported application in this platform assists the system in the context transfer. But, how the interaction in the mobility event between each PoA can have a custom event, to download the video if it's not there?

Thanks!

**Comment by roymx** [comment, marked as answer]
Hello @rubenbc28 ,

Interesting questions.

Firstly could you clarify what _reproducing a streaming video_ means?
Do you mean that _the terminal is streaming a video in the uplink direction to an edge server_?
Do you mean that _the terminal is receiving a video stream in the downlink direction from an edge server?_

To your questions...
1- A PoA does not connect to a streaming server as it is for example a cell tower or WiFi point of access. In edge computing, each PoA can have an edge server associated.
* the _terminal_ that connects to _the network_ via the PoA
* the _application_ on the terminal (e.g. video application) connect to the _edge-application_ (e.g. streaming server) on the associated edge server 
Sorry if I did not get your question right -- please clarify if needed.

2- Here's how MEC021 could be used in your case.
* both edge-servers need to register to MEC021 as the same application type
* when a terminal connects to edge-server-1, then edge-server-1 informs MEC021 that the terminalX is connected (using its IP address)
* then video streaming starts between terminalX and edge-server-1 
* when the terminalX moves to a new PoA, MEC021 detects this and decides that edge-server-2 is better for terminalX -- MEC021 sends a notification to edge-server-1 indicating a state for transfer terminalX is required to edge-server-2
* edge-server-1 then sends a request directly to edge-server-2 (this is a server specific API) indicating an incoming state for terminalX -- the state can include information about the terminalX and should include info such as which video is streamed, the time in the video and possibly some info such as language, subtitles, etc. if available.
* edge-server-2 then does what it needs to do to accommodate terminalX which may include reserve resources, download the video, configure the video time, language, subtitles, etc. 

Hope this helps!

**Reply by ruben-bc** [reply]
Hi Mike,

By _reproducing a streaming video_ I mean that terminal is receiving a video from a streaming server.

The information you have given me helps me to understand it better. Especially for the second question, about which I didn't know if it was possible for edge-server-1 to detect if the edge-server-2 had the video, and download it in case of need.

If I have any other doubt, I will comment on it in this same post.

Thank you!

**Reply by ruben-bc** [reply]
Hi again,

I have another question. In a terminal application it's possible to expose multiple ports. In my case I'm exposing one for the video streaming server and another to open a video player in the browser itself through this application (`<IP>:<Port>`). Without problems I can broadcast a live video with one of the ports that I expose, which maps the RTMP port. And with the other port exposed I can open the video player, but no video appears. The problem I see is that even though multiple ports are exposed through the terminal app, in a edge server only one can be indicated, and then it's not able to do that interconnection between the player and the streaming.

Is there a way to configure various ports to run two different tools together?

Thanks!!

**Reply by roymx** [reply]
Hello @rubenbc28 ,

A terminal, like most compute node, can use multiple ports for IP connectivity.
For example, every time it creates an outgoing TCP connection (ex. connecting to server 1, connecting to server 2, etc.), it is going to use a dynamically allocated port to handle incoming packets associated to the TCP connection.
Likewise, ports can be used by a server inside the terminal to receive connections.

I don't know much about the setup you are describing in your previous message, but here are some of the problems to look out for related to ports:
- NATing/firewall - Is there any NAT between your terminal and the client trying to connect to it?
  - Example:  server on the terminal listens for incoming traffic on Port 12345, network side client shoots traffic to 12345, but the NAT box / firewall sitting in-between does not know anything about 12345 and drops the traffic.
- Well known ports - some application have been designed to use well known ports and can't be configured to use a custom port number

Best way to debug these problems is to do a Wireshark or `tcpdump` capture both on the client and server side -- note that your traffic may make it to the terminal but your application (e.g. browser) might not be hooked to the port you expect
- see if your traffic leaves the client side and if the dest port is as expected
- see if your traffic enters the server side on the expected port

good luck

**Comment by achi07** [comment]
Hei @roymx,

As per your previous comment in this discussion on the MEC 021 use case, can you please explain in details the last two steps which are:

edge-server-1 then sends a request directly to edge-server-2 (this is a server specific API) indicating an incoming state for terminalX -- the state can include information about the terminalX and should include info such as which video is streamed, the time in the video and possibly some info such as language, subtitles, etc. if available.

edge-server-2 then does what it needs to do to accommodate terminalX which may include reserve resources, download the video, configure the video time, language, subtitles, etc.

Also is there any demo that showcase this use case? Thank you so much.

**Comment by roymx** [comment]
Hello @achi07 ,

[Demo3 ](https://interdigitalinc.github.io/AdvantEDGE/docs/usage/usage-demo3/)is the one you want to look at.

To understand the mechanics, see [this webinar](https://www.etsi.org/events/2014-2022-02-webinar-what-s-new-in-the-etsi-mec-sandbox) where we used Demo3 with the MEC Sandbox - it explains the mechanics.

Regarding this ...
_"edge-server-1 then sends a request directly to edge-server-2 (this is a server specific API) indicating an incoming state for terminalX -- the state can include information about the terminalX and should include info such as which video is streamed, the time in the video and possibly some info such as language, subtitles, etc. if available."_

It says that when state transfer happens (e.g. is triggered by the MEC service), the source edge application will contact the target edge application directly to transfer the user state. 
So your application needs to have the following: 
* an API that the source application can call on the target application when the state transfer is triggered
* a definition if what "the state" of your application is 

Regarding this ...
_"edge-server-2 then does what it needs to do to accommodate terminalX which may include reserve resources, download the video, configure the video time, language, subtitles, etc."_

It says that the target application should know what to do with "the state" when it receives it
